### PR TITLE
fix empty blocks

### LINKED ISSUE #127: invalid handling of `parentHash` for skipped blocks during block indexing.
Near blocks might produces some empty blocks which are skipped by the indexer. However there is no check whether skipped parent block hashes (the computed `parentHash` which is the parent of a block previously fetched from the chain) are empty or not.

It requires an extra check in the following line:
https://github.com/aurora-is-near/aurora-relayer/blob/f477c1ff728e24c369a42284b9c25eb015e2e910/src/indexer.ts#L101

### PR DESCRIPTION
**Pull request by 0x3bfc**
closes #127

### CODE REVIEW

File: src/indexer.ts [hunk 1]
```diff
@@ -78,8 +78,37 @@ export class Indexer {
...
-            this.logger.error(error);
-            return; // a skip block, or an unavailable block on a nonarchival node
+            // Handling empty blocks
+            const emptyBlock: EmptyBlock = generateEmptyBlock(
+              blockID as number,
+              contractID.toString(),
+              this.network.chainID,
+              timestamp + 1 // arbitrary timestamp = previous block.timestamp + 1
+            );
+            //if (this.config.debug) console.debug(emptyBlock); // DEBUG
```
Comment (antonpaisov) [inline src/indexer.ts:88]: leftover comment,please remove

File: src/indexer.ts [hunk 2]
```diff
@@ -78,8 +78,37 @@ export class Indexer {
...
+              size: emptyBlock.size,
+              gas_limit: 0, // FIXME: block.gasLimit,
+              gas_used: 0, // FIXME: block.gasUsed,
+              parent_hash: emptyBlock.parentHash,
+              transactions_root: emptyBlock.transactionsRoot,
+              state_root: emptyBlock.stateRoot,
+              receipts_root: emptyBlock.receiptsRoot,
+            });
+
+            //if (this.config.debug) console.debug(query.toString()); // DEBUG
```
Comment (antonpaisov) [inline src/indexer.ts:104]: leftover comment

File: src/indexer.ts [hunk 3]
```diff
@@ -116,7 +146,6 @@ export class Indexer {
         state_root: block.stateRoot,
         receipts_root: block.receiptsRoot,
       });
-
       //if (this.config.debug) console.debug(query.toString()); // DEBUG
```
Comment (antonpaisov) [inline src/indexer.ts:149]: same leftover

**Review by antonpaisov — APPROVED** [review]
lgtm, if well tested lets proceed.
we really need to unblock the graph by tomorrow morning the latest
cc @artob

File: src/indexer.ts (hunk 1 above)
Comment (0x3bfc) [reply src/indexer.ts:88]: This was left for future debugging.

**Review by mfornet — APPROVED** [review]
LGTM

File: src/indexer.ts [hunk 4]
```diff
@@ -66,7 +66,7 @@ export class Indexer {
     //console.debug('indexBlock', blockID); // DEBUG
     this.logger.info({ block: { id: blockID } }, `indexing block #${blockID}`);
     const contractID = AccountID.parse(this.config.engine).unwrap();
-
+    let timestamp = Math.floor(Date.now() / 1000); // init timestamp
```
Comment (artob) [inline src/indexer.ts:69]: This makes indexing nondeterministic: two parallel indexers will not produce the same database entries. That must be fixed. It's better for skip blocks to have no timestamps (it's a nullable column) than to have arbitrary timestamps.

File: src/indexer.ts [hunk 5]
```diff
@@ -78,8 +78,37 @@ export class Indexer {
...
         if (error.startsWith('[-32000] Server error: DB Not Found Error')) {
           if (blockID < currentBlockHeight) {
-            this.logger.error(error);
-            return; // a skip block, or an unavailable block on a nonarchival node
+            // Handling empty blocks
+            const emptyBlock: EmptyBlock = generateEmptyBlock(
+              blockID as number,
+              contractID.toString(),
+              this.network.chainID,
+              timestamp + 1 // arbitrary timestamp = previous block.timestamp + 1
```
Comment (artob) [inline src/indexer.ts:86]: Better for skip blocks to have no timestamps (it's a nullable column) than to have arbitrary timestamps.

File: src/indexer.ts [hunk 6]
```diff
@@ -78,8 +78,37 @@ export class Indexer {
...
+            //if (this.config.debug) console.debug(emptyBlock); // DEBUG
+            const query = sql.insert('block', {
+              chain: this.network.chainID,
+              id: emptyBlock.id,
+              hash: emptyBlock.hash,
+              near_hash: null,
+              timestamp: emptyBlock.timestamp,
+              size: emptyBlock.size,
+              gas_limit: 0, // FIXME: block.gasLimit,
+              gas_used: 0, // FIXME: block.gasUsed,
```
Comment (artob) [inline src/indexer.ts:97]: There is no reason to retain `FIXME` comments here. Just use `emptyBlock.gasLimit` and `emptyBlock.gasUsed`.

File: src/indexer.ts [hunk 7]
```diff
@@ -92,6 +121,7 @@ export class Indexer {
       }
       const block_ = proxy.unwrap();
       const block = block_.getMetadata();
+      timestamp = block.timestamp as number; // it is used only for empty blocks
```
Comment (artob) [inline src/indexer.ts:124]: This can be removed.

File: src/utils.ts [hunk 8]
```diff
@@ -27,3 +42,29 @@ function generateBlockPreImage(
...
+  const hash = computeBlockHash(blockHeight, accountId, chainId);
+
+  const parentHash = computeBlockHash(blockHeight - 1, accountId, chainId);
+
+  return {
+    chain: chainId,
+    id: blockHeight,
+    hash: hash,
+    nearHash: Buffer.alloc(32),
+    timestamp: new Date(timestamp * 1000).toISOString(),
```
Comment (artob) [inline src/utils.ts:61]: Better that skip blocks have no timestamp (it's a nullable column).

**Review by artob — CHANGES_REQUESTED** [review]
The changes as they are would make indexing nondeterministic, so this needs work.

**Review by artob — APPROVED** [review]

**Pull request merged**